rfd/0167-console-service.md

```diff
@@ -0,0 +1,194 @@
+---
+authors: Gabriel Corado (gabriel.oliveira@goteleport.com)
+state: draft
+---
+
+# RFD 0167 - Console service
+
+## Required Approvals
+
+* Engineering: @r0mant && @greedy52
+* Security: @reedloden || @jentfoo
+
+## What
+This RFD proposes a mechanism for Teleport administrators to switch log-level
+and enable profile consumption without restarting the instance.
+
+## Why
+During incidents in production environments, debug logs and diagnosis profiles
+(such as CPU and Memory allocation) are often unavailable for timely
+troubleshooting. To enable those, administrators need to restart the instance,
+potentially disrupting the application state and obscuring the issue they're
+trying to solve.
+
+## Details
+
+A separate service will be introduced specifically for troubleshooting
+purposes. This will ensure the current diagnosis service is kept as is, not
+breaking current integrations or introducing behavior changes. The new service
+will also always be available (contrasting with the diagnosis service, which
+is optional and disabled by default), so in scenarios where users need to use
+it, they don't need to restart Teleport.
+
+In addition, the new service will listen into a Unix socket instead of TCP. This
+will make discoverability easier and discourage external usage, as the API is
+designed for internal usage.
+
+The `pprof` endpoints on diagnostics will still be available for existent
+integrations, such as the usage of
+`go tool pprof http://diag-addr/debug/pprof/profile`.
+
+### New service
+
+Teleport will start listening using a Unix socket located at
+`/tmp/teleport-diag-<node-id>.sock`.
+
+Having the Node ID on the socket name will also cover scenarios where multiple
+instances of running on the same machine exist. In this case, the consumers can
+rely on the Teleport configuration to locate the data directory and retrieve
+the ID.
```

Comment:
I'm still testing this. I can tell from now that the [listener initialization process](https://github.com/gravitational/teleport/blob/ddad68338ba108762c7a5f240d159a9a64c25a10/lib/service/signals.go#L260-L273) will work the same way we're doing for TCP listeners (as it will use the code, expect that the `network` will be different), which takes into account [imported file descriptors](https://github.com/gravitational/teleport/blob/ddad68338ba108762c7a5f240d159a9a64c25a10/lib/service/signals.go#L290-L308).

Reply:
After testing, I just confirmed that the file descriptors are imported to the new child process correctly (with the same flow as other file descriptors/listeners).

Reply:
I would 100% not move the listener to the child - you can just unlink the path and reuse it with a new listener, it's not like listening on the same TCP port.